### PR TITLE
feat: pin cert-manager image to sha256 checksum

### PR DESCRIPTION
**Pull request by elchead**
### Context

Pin the cert-manager related images to their sha256 checksum. 

How I tested it:
upgraded a GCP, v2.13 cluster with
```
./constellation apply --debug --skip-phases infrastructure,k8s,image --force
```

### Proposed change(s)

-

### Additional info

-  [AB#3612](https://dev.azure.com/Edgeless/ae37573d-ccde-4af2-ab1e-001e587197d1/_workitems/edit/3612)
Had to rebase and cherry pick to be able to test it locally.

### Checklist

- [ ] Update [docs](https://github.com/edgelesssys/constellation/tree/main/docs)
- [ ] Add labels (e.g., for changelog category)
- [ ] Is PR title adequate for changelog?
- [ ] Link to Milestone

### CODE REVIEW

File: internal/constellation/helm/generateCertManager.sh [hunk 1]
```diff
@@ -18,4 +18,20 @@ helm pull cert-manager \
   --untar \
   --untardir "charts"
 
+echo "Pinning cert-manager images..."
```
Comment (3u13r) [inline internal/constellation/helm/generateCertManager.sh:21]: I assume that yq already formats the helm charts. Since those values are exposed as helm values, I favor just setting them like all the other helm values.

**Review by 3u13r — DISMISSED** [review]

File: internal/constellation/helm/generateCertManager.sh (hunk 1 above)
Comment (elchead) [reply internal/constellation/helm/generateCertManager.sh:21]: I don't like that approach because we then separate the helm version (set in the script) from the according hashes.

Comment (daniel-weisse) [reply internal/constellation/helm/generateCertManager.sh:21]: Setting the hashes later (i.e. "like all the other helm values") would allow us do update the hashes using renovate.
We already do this for the cluster-autoscaler and cloud-controller-manager images, and some others.
See our [versions.go](https://github.com/edgelesssys/constellation/blob/a1dbd13f9577e23d99012416ab1b97c83ecfdedb/internal/versions/versions.go) file.
Not 100% how/if we want that for charts we pull in as third party dependencies since the versions should usually not be updated for one version of a chart, but it could make things easier since we wouldn't need to manually recalculate the image hashes every time we update the chart

Comment (elchead) [reply internal/constellation/helm/generateCertManager.sh:21]: I assume @burgerdev motivation was to protect against a malicious version overwrite in the registry?
I agree that it would be good to use renovate.

File: internal/constellation/helm/generateCertManager.sh [hunk 2]
```diff
@@ -18,4 +18,21 @@ helm pull cert-manager \
...
+# Manually copy the sha256 digest from the corresponding repository:
+# quay.io/jetstack/cert-manager-controller
+yq -i '(.image.digest) = "sha256:fb2546fe51e49206dbf72bb0d6f909a0018eda0c2b024547b03d3f3d604e4c5e"' charts/cert-manager/values.yaml
```
Comment (burgerdev) [inline internal/constellation/helm/generateCertManager.sh:24]: We could auto-generate the hashes along the lines of

```bash
curl -L -H "Accept: application/vnd.docker.distribution.manifest.v2+json" https://quay.io/v2/jetstack/${component}/manifests/v${version} | sha256sum
```

File: internal/constellation/helm/generateCertManager.sh (hunk 1 above)
Comment (burgerdev) [reply internal/constellation/helm/generateCertManager.sh:21]: I'm facing more or less the same issue in my other hash pinning attempts, and this is the approach that I feel most comfortable with so far:

1. We want to specify a (chart|k8s|.*) version `$SOMEWHERE`, which ideally would be renovated.
2. As a consequence of that version choice, a list of required container images emerges.
3. We automatically derive the image hashes for those
4. and put them `$SOMEWHERE`.

The place for (1) is currently this script. I think that it would make sense to do (2-4) here as well, at least in the short term. If we can come up with 

* a place for storing (1) 
  - _globally_
* a generator with input (1) and output (2) 
  - _per use case_
* a generator with input (2) and output (3) 
  - _globally_
* a generator with input (3) 
  - _for a couple of different use cases_, e.g. `versions.go`, `values.go`

life would be a lot easier, but I think that's outside the scope of this change. However, I do think it's worthwile considering.

Long story short, +1 to keeping the chart version and the image hashes close together.

**Comment by daniel-weisse** [comment]
Since finding the hashes to pin for images not controlled by us seems to be a recurring issue, and is not easily solved:
How about just keeping it simple for this PR, and then open an RFC to discuss options to have the hash finding centralized at some place?

File: internal/constellation/helm/generateCertManager.sh (hunk 2 above)
Comment (elchead) [reply internal/constellation/helm/generateCertManager.sh]: thanks, implemented that now

File: internal/constellation/helm/generateCertManager.sh [hunk 3]
```diff
@@ -12,10 +12,51 @@ function cleanup {
...
+  if [ "$has_error" = "true" ]; then
+      echo "Errors attribute found. Exiting with status 1."
+      exit 1
+  fi
+
+  hash=$(curl -s -L -H "Accept: application/vnd.docker.distribution.manifest.v2+json" "$url" | sha256sum | awk '{print $1}')
+
+  set +e  # Disable the 'set -e' option to avoid exiting on subsequent commands
+
+  echo "$hash"
```
Comment (burgerdev) [inline internal/constellation/helm/generateCertManager.sh:41]: `set -e` is inherited from line 3, and `set +e` resets globally. I don't think you need to change this at all.

```suggestion
  has_error=$(curl -s -L -H "Accept: application/vnd.docker.distribution.manifest.v2+json" "$url" | jq 'has("errors")')

  # Check if 'errors' attribute exists
  if [ "$has_error" = "true" ]; then
      echo "Errors attribute found. Exiting with status 1."
      exit 1
  fi

  curl -s -L -H "Accept: application/vnd.docker.distribution.manifest.v2+json" "$url" | sha256sum | awk '{print $1}'
```

File: internal/constellation/helm/generateCertManager.sh [hunk 4]
```diff
@@ -12,10 +12,51 @@ function cleanup {
 
 trap cleanup EXIT
 
+version="1.12.6"
```
Comment (burgerdev) [inline internal/constellation/helm/generateCertManager.sh:15]: This variable needs to be used in the cleanup function at line 10.

**Review by burgerdev — CHANGES_REQUESTED** [review]

**Review by burgerdev — APPROVED** [review]

File: internal/constellation/helm/generateCertManager.sh [hunk 5]
```diff
@@ -5,17 +5,51 @@ set -o errtrace
...
+  local hash
+  has_error=$(curl -s -L -H "Accept: application/vnd.docker.distribution.manifest.v2+json" "$url" | jq 'has("errors")')
+
+  # Check if 'errors' attribute exists
+  if [ "$has_error" = "true" ]; then
+    echo "Errors attribute found. Exiting with status 1."
+    exit 1
+  fi
+
+  curl -s -L -H "Accept: application/vnd.docker.distribution.manifest.v2+json" "$url" | sha256sum | awk '{print $1}'
```
Comment (3u13r) [inline internal/constellation/helm/generateCertManager.sh:34]:
```suggestion
  curl -fsSL -H "Accept: application/vnd.docker.distribution.manifest.v2+json" "$url" | sha256sum | awk '{print $1}'
```

File: internal/constellation/helm/generateCertManager.sh [hunk 6]
```diff
@@ -5,17 +5,49 @@ set -o errtrace
...
 function cleanup {
-  rm -r "charts/cert-manager/README.md" "charts/cert-manager-v1.12.6.tgz"
+  rm -r "charts/cert-manager/README.md" "charts/cert-manager-v${version}.tgz"
 }
 
 trap cleanup EXIT
 
 helm pull cert-manager \
-  --version 1.12.6 \
+  --version ${version} \
```
Comment (3u13r) [inline internal/constellation/helm/generateCertManager.sh:17]:
```suggestion
  --version "${version}" \
```

File: internal/constellation/helm/generateCertManager.sh [hunk 7]
```diff
@@ -5,17 +5,49 @@ set -o errtrace
...
   --untardir "charts"
 
+get_sha256_hash() {
+  local component="$1"
+  local url="https://quay.io/v2/jetstack/${component}/manifests/v${version}"
+  local hash
+  has_error=$(curl -s -L -H "Accept: application/vnd.docker.distribution.manifest.v2+json" "$url" | jq 'has("errors")')
+
+  # Check if 'errors' attribute exists
+  if [ "$has_error" = "true" ]; then
```
Comment (3u13r) [inline internal/constellation/helm/generateCertManager.sh:29]:
```suggestion
  if [[ "$has_error" = "true" ]]; then
```

File: internal/constellation/helm/generateCertManager.sh [hunk 8]
```diff
@@ -5,17 +5,49 @@ set -o errtrace
...
+yq eval -i '.webhook.image.digest = "sha256:'"$v"'"' charts/cert-manager/values.yaml
+
+v=$(get_sha256_hash "cert-manager-cainjector")
+yq eval -i '.cainjector.image.digest = "sha256:'"$v"'"' charts/cert-manager/values.yaml
+
+v=$(get_sha256_hash "cert-manager-acmesolver")
+yq eval -i '.acmesolver.image.digest = "sha256:'"$v"'"' charts/cert-manager/values.yaml
+
+v=$(get_sha256_hash "cert-manager-ctl")
+yq eval -i '.startupapicheck.image.digest = "sha256:'"$v"'"' charts/cert-manager/values.yaml
```
Comment (3u13r) [inline internal/constellation/helm/generateCertManager.sh:51]:
```suggestion
yq eval -i ".startupapicheck.image.digest = \"sha256:${v}\"" charts/cert-manager/values.yaml
```
Also see [SC2250](https://www.shellcheck.net/wiki/SC2250) everywhere a variable is referenced.

**Review by 3u13r — DISMISSED** [review]
